fix: cast node.size to IntegerLiteral for qubit register size

rmshaffer opened the pull request:
*Issue #, if available:*
#240 

*Description of changes:*
Starting with the changes from #258 and performing the additional work required to merge to `main`.

The biggest outstanding items are:
- [ ] Refactor the new logic in `functions.py` into a separate helper function, for readability. See [this comment thread](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/258#discussion_r1643096457).
- [ ] Expand unit tests to ensure correctness of implementation. See [this comment thread](https://github.com/amazon-braket/amazon-braket-default-simulator-python/pull/258#discussion_r1643097868).
- [ ] Discuss among maintainers to ensure that the new behavior is actually in line with how we want to support OpenQASM in the default simulator (since the OpenQASM spec isn't entirely clear on all of this).

*Testing done:*
`tox`

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md) doc
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/amazon-braket/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#commit-your-change)
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/amazon-braket/amazon-braket-default-simulator-python/blob/main/README.md) and [API docs](https://github.com/amazon-braket/amazon-braket-default-simulator-python/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

EuGig committed:
fix: cast node.size to IntegerLiteral for qubit register size (#258)

Fixes bug [#240] by casting node.size into IntegerLiteral

EuGig commented:
@rmshaffer, @DanBlackwell
Maybe we should rethink the following case test:
```
bit[3] b;
qubit["10"] r1;
h r1["01" << "01"];
``` 
In this regard the spec explicitly states that shifting, either right or left, is supported by an unsigned integer (https://openqasm.com/language/classical.html#classical-bits-and-registers).
Shifting by a `bitstring` causes a strange behaviour in case its length does not match the length of the specified `qbuit` array:
```
bit[3] b;
qubit[3] r1;
h r1["01" << "01"];
IndexError: boolean index did not match indexed array along dimension 0; dimension is 3 but corresponding boolean dimension is 2
```
Since we are enabling registers sizes to be specified also as `bitstring`, maybe we should raise an error if the LHS is also a `bitstring`.
This influences the refactoring since, it pertains code regarding shifting operations.

DanBlackwell commented:
Maybe slightly related, but I found the following [here](https://docs.quantum.ibm.com/build/qasm-feature-table) the other day:

<img width="796" alt="image" src="https://github.com/amazon-braket/amazon-braket-default-simulator-python/assets/10620369/b369032b-ef78-46f0-9198-d4bcd9322fbb">

I'm not sure whether this is implying that it is valid to cast an array of bits into int and vice versa; but in that case I think that when applying any ops (`+ - << >> | & ^` etc) to an 'array of bits' (which a 'bit string literal' is afaik) the best approach is to cast to 'int' rather than reimplement them (and cast back if required). I'm not sure how integer over- and underflows are meant to be handled though.

rmshaffer committed:
Merge branch 'main' into feature/integer-literal-cast

speller26 committed:
Merge branch 'main' into feature/integer-literal-cast

speller26 committed:
Merge branch 'main' into feature/integer-literal-cast

rmshaffer commented:
Closing this PR as it has not been a priority to finish this functionality. If anyone is affected by this and wants to contribute a fresh PR which addresses the points raised in the comment threads above, you are more than welcome.
